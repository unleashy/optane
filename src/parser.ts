function snakeToCamel(snake: string): string {
  return snake.replaceAll(
    /([^-])-([^-])/g,
    (_, l: string, r: string) => `${l}${r.toUpperCase()}`,
  );
}

export type Element =
  | { type: "option"; name: string }
  | { type: "free"; value: string }
  | { type: "end-options" };

export type Result = { elements: Element[]; errors: string[] };

export function parse(argv: string[]): Result {
  let cleanArgv = argv
    .map((value) => value.trim())
    .filter((value) => value.length > 0);

  let errors: string[] = [];
  let optionsEnded = false;
  let elements: Element[] = cleanArgv.flatMap((arg): Element[] => {
    if (!optionsEnded) {
      if (arg === "--") {
        optionsEnded = true;
        return [{ type: "end-options" }];
      }

      if (arg.startsWith("--")) {
        let name = snakeToCamel(arg.slice(2));

        return [{ type: "option", name }];
      } else if (arg.startsWith("-") && arg !== "-") {
        let name = arg.slice(1);
        if (name.length > 1) {
          errors.push(
            `-${name}: short options must only use a single character`,
          );
          return [];
        }

        return [{ type: "option", name }];
      }
    }

    return [{ type: "free", value: arg }];
  });

  return { elements, errors };
}

////////////////////////////////////////////////////////////////////////////////

if (import.meta.vitest) {
  const { test, expect } = import.meta.vitest;

  test("empty args", () => {
    expect(parse([])).toMatchSnapshot();
  });

  test("one positional arg", () => {
    expect(parse(["foo"])).toMatchSnapshot();
  });

  test("multiple positional args", () => {
    expect(parse(["foo", "bar", " baz "])).toMatchSnapshot();
  });

  test("empty strings are ignored", () => {
    expect(parse(["", "a", "   ", "b", "\n", "c"])).toMatchSnapshot();
  });

  test("one short option", () => {
    expect(parse(["-a"])).toMatchSnapshot();
  });

  test("multiple short options", () => {
    expect(parse(["-x", "-y", "-z"])).toMatchSnapshot();
  });

  test("short options must be a single character", () => {
    expect(parse(["-foo"])).toMatchSnapshot();
  });

  test("one long option", () => {
    expect(parse(["--foo"])).toMatchSnapshot();
  });

  test("multiple long options", () => {
    expect(parse(["--bread", "--eggs", "--ham"])).toMatchSnapshot();
  });

  test("mixed arguments and options", () => {
    expect(
      parse(["rice", "--sausage", "french", "-f", "ries"]),
    ).toMatchSnapshot();
  });

  test("snake-case becomes camelCase", () => {
    expect(
      parse([
        "--a-b",
        "--foo-bar",
        "--is-the-thing-on",
        "--a",
        "--a--b--c",
        "--a-",
        "---b",
      ]),
    ).toMatchSnapshot();
  });

  test("- is not interpreted as an option", () => {
    expect(parse(["-"])).toMatchSnapshot();
  });

  test("-- is interpreted as the end of options", () => {
    expect(parse(["--", "-a"])).toMatchSnapshot();
  });
}
