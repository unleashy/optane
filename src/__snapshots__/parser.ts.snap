// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`empty args 1`] = `
{
  "elements": [],
  "errors": [],
}
`;

exports[`empty strings are ignored 1`] = `
{
  "elements": [
    {
      "type": "free",
      "value": "a",
    },
    {
      "type": "free",
      "value": "b",
    },
    {
      "type": "free",
      "value": "c",
    },
  ],
  "errors": [],
}
`;

exports[`mixed arguments and options 1`] = `
{
  "elements": [
    {
      "type": "free",
      "value": "rice",
    },
    {
      "isShort": false,
      "name": "sausage",
      "type": "option",
    },
    {
      "type": "free",
      "value": "french",
    },
    {
      "isShort": true,
      "name": "f",
      "type": "option",
    },
    {
      "type": "free",
      "value": "ries",
    },
  ],
  "errors": [],
}
`;

exports[`multiple long options 1`] = `
{
  "elements": [
    {
      "isShort": false,
      "name": "bread",
      "type": "option",
    },
    {
      "isShort": false,
      "name": "eggs",
      "type": "option",
    },
    {
      "isShort": false,
      "name": "ham",
      "type": "option",
    },
  ],
  "errors": [],
}
`;

exports[`multiple positional args 1`] = `
{
  "elements": [
    {
      "type": "free",
      "value": "foo",
    },
    {
      "type": "free",
      "value": "bar",
    },
    {
      "type": "free",
      "value": "baz",
    },
  ],
  "errors": [],
}
`;

exports[`multiple short options 1`] = `
{
  "elements": [
    {
      "isShort": true,
      "name": "x",
      "type": "option",
    },
    {
      "isShort": true,
      "name": "y",
      "type": "option",
    },
    {
      "isShort": true,
      "name": "z",
      "type": "option",
    },
  ],
  "errors": [],
}
`;

exports[`one long option 1`] = `
{
  "elements": [
    {
      "isShort": false,
      "name": "foo",
      "type": "option",
    },
  ],
  "errors": [],
}
`;

exports[`one positional arg 1`] = `
{
  "elements": [
    {
      "type": "free",
      "value": "foo",
    },
  ],
  "errors": [],
}
`;

exports[`one short option 1`] = `
{
  "elements": [
    {
      "isShort": true,
      "name": "a",
      "type": "option",
    },
  ],
  "errors": [],
}
`;

exports[`short options must be a single character 1`] = `
{
  "elements": [],
  "errors": [
    "-foo: short options must only use a single character",
  ],
}
`;

exports[`snake-case becomes camelCase 1`] = `
{
  "elements": [
    {
      "isShort": false,
      "name": "aB",
      "type": "option",
    },
    {
      "isShort": false,
      "name": "fooBar",
      "type": "option",
    },
    {
      "isShort": false,
      "name": "isTheThingOn",
      "type": "option",
    },
    {
      "isShort": false,
      "name": "a",
      "type": "option",
    },
    {
      "isShort": false,
      "name": "a--b--c",
      "type": "option",
    },
    {
      "isShort": false,
      "name": "a-",
      "type": "option",
    },
    {
      "isShort": false,
      "name": "-b",
      "type": "option",
    },
  ],
  "errors": [],
}
`;
